src/backend/aspen/api/views/auspice.py

```diff
@@ -0,0 +1,204 @@
+    that as the key."""
+    for node in tree:
+        name = node["name"]
+        renamed_value = name_map.get(name, None)
+        if renamed_value is not None:
+            # we found the replacement value! first, save the old value if the caller
+            # requested.
+            if save_key is not None:
+                node[save_key] = name
+            node["name"] = renamed_value
+        if "children" in node:
+            _rename_nodes_on_tree(node["children"], name_map, save_key)
+    return tree
+
+
+async def _verify_phylo_tree_access(
+    db: AsyncSession, user: User, phylo_tree_id: int, load_samples: bool = False
+) -> Tuple[bool, Optional[PhyloTree]]:
+    tree_query = sa.select(PhyloTree).join(PhyloRun)
+    if load_samples:
+        tree_query = tree_query.options(selectinload(PhyloTree.constituent_samples))
+    authz_tree_query = authz_phylo_tree_filters(tree_query, user, set([phylo_tree_id]))
+    authz_tree_query_result = await db.execute(authz_tree_query)
+    phylo_tree: PhyloTree
+    try:
+        phylo_tree = authz_tree_query_result.scalars().unique().one()
+    except sa.exc.NoResultFound:  # type: ignore
+        return False, None
+    return True, phylo_tree
+
+
+def _sample_filter(sample: Sample, can_see_pi_group_ids: Set[int], system_admin: bool):
+    if system_admin:
+        return True
+    return sample.submitting_group_id in can_see_pi_group_ids
+
+
+async def _get_and_filter_phylo_tree(
+    db: AsyncSession, user: User, phylo_tree_id: int
+) -> dict:
+    authorized, phylo_tree_result = await _verify_phylo_tree_access(
+        db, user, phylo_tree_id, load_samples=True
+    )
+    if not authorized or not phylo_tree_result:
+        raise ex.BadRequestException("No phylo run found for auspice request")
+    phylo_tree: PhyloTree = phylo_tree_result
+
+    can_see_pi_group_ids: Set[int] = {user.group_id}
+    can_see_pi_group_ids.update(
+        {
+            can_see.owner_group_id
+            for can_see in user.group.can_see
+            if can_see.data_type == DataType.PRIVATE_IDENTIFIERS
+        }
+    )
+
+    identifier_map: Mapping[str, str] = {
+        sample.public_identifier.replace("hCoV-19/", ""): sample.private_identifier
+        for sample in phylo_tree.constituent_samples
+        if _sample_filter(sample, can_see_pi_group_ids, user.system_admin)
+    }
+
+    s3 = boto3.resource(
+        "s3",
+        endpoint_url=os.getenv("BOTO_ENDPOINT_URL") or None,
+        config=boto3.session.Config(signature_version="s3v4"),
+    )
+
+    data = (
+        s3.Bucket(phylo_tree.s3_bucket).Object(phylo_tree.s3_key).get()["Body"].read()
+    )
+    json_data = json.loads(data)
+
+    json_data = _rename_nodes_on_tree([json_data["tree"]], identifier_map, "GISAID_ID")
+
+    return json_data
+
+
+# TODO: Convert to POST request
+@router.post("/generate")
+async def generate_auspice_string(
+    request: Request,
```

Comment:
this request can be a GenerateAuspiceMagicLinkRequest and fastapi/pydantic will validate the input automatically for you

Reply:
True, but we need access to `request.url.netloc` to generate the url we are returning. Is there a way to make a pydantic schema preserve that information?

Reply:
Fair point. It might be possible to do it other ways but this seems fine for now.